agents/self_evolving_car/train.py

```diff
@@ -0,0 +1,163 @@
+from examples.levi.output_formatter import LeviOutputFormatter
+import sys, os
+
+
+class PythonExample(BaseAgent):
+    def __init__(self, name, team, index):
+        super().__init__(name, team, index)
+
+        import torch
+        sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))  # this is for separate process imports
+        from examples.levi.torch_model import SymmetricModel
+        self.Model = SymmetricModel
+        self.torch = torch
+        self.controller_state = SimpleControllerState()
+        self.frame = 0  # frame counter for timed reset
+        self.brain = 0  # bot counter for generation reset
+        self.pop = 10  # population for bot looping
+        self.num_best = 5
+        self.gen = 0
+        self.pos = 0
+        self.max_frames = 5000
+        self.bot_list = [self.Model() for _ in range(self.pop)]  # list of Individual() objects
+        self.bot_list[-self.num_best:] = [self.Model()] * self.num_best  # make sure last bots are the same
+        self.bot_fitness = [0] * self.pop
+        self.fittest = None  # fittest object
+        self.mut_rate = 0.2  # mutation rate
+        self.distance_to_ball = [math.inf] * self.max_frames  # set high for easy minimum
+        self.input_formatter = LeviInputFormatter(team, index)
+        self.output_formatter = LeviOutputFormatter(index)
+
+    def initialize_agent(self):
+        self.reset()  # reset at start
+
+    def get_output(self, packet: GameTickPacket) -> SimpleControllerState:
+        # NEURAL NET INPUTS
+        inputs = self.input_formatter.create_input_array([packet])
+        inputs = [self.torch.from_numpy(x).float() for x in inputs]
+
+        my_car = packet.game_cars[self.index]
+        distance_to_ball_x = packet.game_ball.physics.location.x - my_car.physics.location.x
+        distance_to_ball_y = packet.game_ball.physics.location.y - my_car.physics.location.y
+        distance_to_ball_z = packet.game_ball.physics.location.z - my_car.physics.location.z
+        self.distance_to_ball[self.frame] = math.sqrt(distance_to_ball_x**2 + distance_to_ball_y**2 + distance_to_ball_z**2)
+
+        # RENDER RESULTS
+        action_display = "GEN: "+str(self.gen+1)+" | BOT: "+str(self.brain)#+" \nTURN: "+str(self.botList[self.brain].Nodes[1].node(self.distance_to_ball_x,self.distance_to_ball_y,self.distance_to_ball_z,4,5,6,7))+" \nBOOST: "+str(self.botList[self.brain].Nodes[2].node(self.distance_to_ball_x,self.distance_to_ball_y,self.distance_to_ball_z,8,9,10,11))+" \nTHROTTLE: "+str(self.botList[self.brain].Nodes(self.distance_to_ball_x,self.distance_to_ball_y,self.distance_to_ball_z,0,1,2,3))
+
+        draw_debug(self.renderer, my_car, packet.game_ball, action_display)
+
+        # # STOP EVOLVING WHEN THE BALL IS TOUCHED
+        # if packet.game_ball.latest_touch.player_name == "Self-Evolving-Car":
+        #     self.mut_rate = 0
+
+        # GAME STATE
+        car_state = CarState(boost_amount=100)
```

Comment:
look into using domnomnom framework for setting up scenarios.  it might be more extensible for long term changes like having multiple preset ball locations